networkx/algorithms/approximation/density.py

```diff
@@ -0,0 +1,166 @@
+        heap = nx.utils.BinaryHeap()
+
+        # Calculate the current weighted degrees (load + degree of node) and add it to the heap.
+        weighted_degrees = {u: loads[u] + d for u, d in G.degree()}
+        for u in G.nodes:
+            heap.insert(u, weighted_degrees[u])
+
+        # Copy the graph structure for this iteration
+        G_iter = G.copy()
+
+        while G_iter.number_of_nodes() >= 1:
+            current_density = (
+                G_iter.number_of_edges() / G_iter.number_of_nodes()
+            )  # Current density of G_iter
+
+            if current_density > best_density:
+                best_density = current_density
+                best_subgraph = set(G_iter.nodes)
+
+            u, _ = heap.pop()  # Pick node with minimum current weighted degree
+
+            loads[u] += G_iter.degree[u]  # Add to load of u
+
+            # Subtract one edge from all of u's neighbors, and update their weighted degree
+            for v in G_iter.neighbors(u):
+                weighted_degrees[v] -= 1
+                heap.insert(v, weighted_degrees[v])
+
+            # Delete u
+            G_iter.remove_node(u)
+
+    return (best_density, best_subgraph)
+
+
+ALGORITHMS = {"greedy++": _greedy_plus_plus}
+
+
+@nx.utils.not_implemented_for("directed")
+@nx.utils.not_implemented_for("multigraph")
+@nx._dispatchable
+def densest_subgraph(G, iterations=1, *, method="greedy++"):
+    r"""Returns an approximate densest subgraph for a graph `G`.
+
+    This function runs an iterative algorithm to find
+    the densest subgraph, and returns both the density and the subgraph.
+    For a discussion on the notion of density used and the different
+    algorithms available on networkx, please see the Notes section below.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        Undirected graph.
+
+    iterations : int, optional (default=1)
+        Number of iterations to use for the iterative algorithm.
+        Can be specified positionally or as a keyword argument.
+
+    method : string, optional (default='greedy++')
+        The algorithm to use to approximate the densest subgraph.
+        Supported options: 'greedy++'.
+        Must be specified as a keyword argument. Other inputs produce a ValueError.
+
+    Returns
+    -------
+    d : float
+        The density of the approximate subgraph found.
+
+    S : set
+        The subset of nodes defining the approximate densest subgraph.
+
+    Examples
+    --------
+    >>> G = nx.star_graph(4)
+    >>> nx.approximation.densest_subgraph(G, iterations=1)
+    (0.8, {0, 1, 2, 3, 4})
+
+    Notes
+    -----
+    For a subset of the nodes of $G$, $S \subseteq V(G)$, define
+    $E(S) = \{ (u,v) : (u,v)\in E(G), u\in S, v\in S \}$ as the set of
+    edges with both endpoints in $S$. The density of $S$ is defined as
+    $E(S)/|S|$, the ratio between the edges in the subgraph $G[S]$ and
+    the number of nodes in that subgraph. The densest subgraph problem
```

Comment:
Maybe put the final sentence in this paragraph first. And introduce the abbreviation DSG here so you can use it later. (like:  `The densest subgraph problem (DSG) asks...`
I think this description could be simplified by replacing the set notation with words like `induced subgraph`. But mostly I think the term "density" is not correctly defined here -- at least it doesn't match the standard graph theory definition. ([wikipedia](https://en.wikipedia.org/wiki/Dense_graph))